app/src/main/java/io/neurolab/model/FFTPreprocessor.java

```diff
@@ -0,0 +1,142 @@
+package io.neurolab.model;
+
+import java.util.Iterator;
+
+import edu.emory.mathcs.jtransforms.fft.DoubleFFT_1D;
+import io.neurolab.NeuroSettings;
+import io.neurolab.tools.WindowFunction;
+
+public class FFTPreprocessor implements Task {
+    private NeuroSettings bciSettings;
+    private DoubleFFT_1D fft;
+    protected DefaultFFTData fftData;
+    private double[] filterWindow;
+
+    private int numFIRFilterSamples = 10;
+    private double avgValCount = 5;
+
+    public FFTPreprocessor(DefaultFFTData fftData, NeuroSettings bciSettings) {
+        this.bciSettings = bciSettings;
+        this.fftData = fftData;
+        init();
+    }
+
+    public void setNumFIRFilterSamples(int numFIRFilterSamples) {
+        this.numFIRFilterSamples = numFIRFilterSamples;
+    }
+
+    public int getNumFIRFilterSamples() {
+        return numFIRFilterSamples;
+    }
+
+    public void init() {
+        this.fft = new DoubleFFT_1D(bciSettings.getSamplesPerSecond());
+        this.fftData.init(bciSettings.getSamplesPerSecond(), bciSettings.getBins(), bciSettings.getNumChannels());
+        System.out.println("window size: " + this.fftData.getWindowSize() + " / samples per second: " + bciSettings.getSamplesPerSecond());
+        this.filterWindow = WindowFunction.generate(this.fftData.getWindowSize(), WindowFunction.FunctionType.BLACKMAN);
+    }
+
+    public DefaultFFTData getFFTData() {
+        return fftData;
+    }
+
+    private void nullifyFft() {
+        fft = null;
+    }
+
+    @Override
+    public void run() {
+
+        Iterator<double[]> currentDataIterator = new Iterator<double[]>() {
+            @Override
+            public boolean hasNext() {
+                return false;
+            }
+
+            @Override
+            public double[] next() {
+                return new double[0];
+            }
+        };
+
+        int s = 0;
+        double diff = 0;
+        int fftDataLocal = fftData.getNumChannels();
+        fftDataLocal = bciSettings.getNumChannels();
+
+        boolean notBrainWaves[];
+        notBrainWaves = new boolean[bciSettings.getNumChannels()];
+        while (currentDataIterator.hasNext()) {
```

Comment:
This loop would never get entered as hasNext( ) method of the currentDataIterator is set to false above. The correct way would be to use:
 `Iterator<double[]> currentDataIterator = bciSettings.getCurrentData().iterator();`
For which you have to add ConcurrentLinkedDeque<double[]> object in NeuroSettings class (so that it can be accessed and used appropriately)

Reply:
Yeah I figured too.